plugins/nvme/fu-plugin-nvme.c

```diff
@@ -0,0 +1,289 @@
+/*
+ * Copyright (C) 2018 Richard Hughes <richard@hughsie.com>
+ *
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+
+#include "config.h"
+
+#include <gudev/gudev.h>
+#include <json-glib/json-glib.h>
+#include <string.h>
+
+#include "fu-plugin.h"
+#include "fu-plugin-vfuncs.h"
+
+struct FuPluginData {
+	gchar			*nvme_fn;
+	GUdevClient		*gudev_client;
+};
+
+static gboolean
+fu_plugin_nvme_add_devfile (FuPlugin *plugin, const gchar *devfile, GError **error)
+{
+	FuPluginData *data = fu_plugin_get_data (plugin);
+	JsonNode *node_root;
+	JsonObject *obj;
+	const gchar *nvme_fr;
+	const gchar *nvme_mn;
+	gint64 nvme_frmw;
+	g_autofree gchar *nvme_mn_fixed = NULL;
+	g_autofree gchar *stdout = NULL;
+	g_auto(GStrv) nvme_mn_split = NULL;
+	g_autoptr(FuDevice) dev = fu_device_new ();
+	g_autoptr(JsonParser) parser = json_parser_new ();
+	const gchar *argv[] = { data->nvme_fn, "id-ctrl", devfile,
+				"--output-format=json", NULL };
+
+	/* get JSON data */
+	if (!g_spawn_sync (NULL, (gchar **) argv, NULL, G_SPAWN_DEFAULT,
+			   NULL, NULL, &stdout, NULL, NULL, error))
+		return FALSE;
+	if (!json_parser_load_from_data (parser, stdout, -1, error)) {
+		g_prefix_error (error, "failed to parse node: ");
+		return FALSE;
+	}
+	node_root = json_parser_get_root (parser);
+	if (node_root == NULL) {
+		g_set_error_literal (error,
+				     FWUPD_ERROR,
+				     FWUPD_ERROR_INVALID_FILE,
+				     "no root node");
+		return FALSE;
+	}
+	obj = json_node_get_object (node_root);
+	if (obj == NULL) {
+		g_set_error_literal (error,
+				     FWUPD_ERROR,
+				     FWUPD_ERROR_INVALID_FILE,
+				     "no object");
+		return FALSE;
+	}
+
+	/* create device */
+	fu_device_add_flag (dev, FWUPD_DEVICE_FLAG_INTERNAL);
+	fu_device_add_flag (dev, FWUPD_DEVICE_FLAG_UPDATABLE);
+	fu_device_add_flag (dev, FWUPD_DEVICE_FLAG_REQUIRE_AC);
```

Comment:
This collection of flags I don't think is necessarily true.  Since NVMe drives are PCIe device, it's certainly possible to have them in a Thunderbolt device or thunderbolt PCIe enclosure.  To really prove they are internal/require-ac you'd have to look at the topology to see if they go through a thunderbolt bridge.

I don't think that's worth it though.

Reply:
Are there any good heuristics for the topology walking? I assume we just need to look at the bridges and root controllers.

Reply:
Maybe @gicmo or @YehezkelShB have some ideas here.  My beset guess for a heuristic would be to match the thunderbolt NHI uevent to find it's parent and then see if the parent is in common.